src/components/ArmyBuildPane.jsx

```diff
@@ -17,30 +17,42 @@ const Container = styled.div`
 `;
 
 const ArmyBuildPane = () => {
+
   const [unitList, setUnitList] = useState({});
   const [listIndex, setListIndex] = useState(1);
 
   const addUnit = unit => {
+    unit.selected_upgrades = {};
     setUnitList({
       ...unitList,
       [listIndex]: unit
     });
     setListIndex(listIndex + 1);
   };
+
   const removeUnit = index => {
-    const {[index]: removedValue, ...remaining} = unitList;
-    setUnitList(
-      remaining
-    );
+    const { [index]: removedValue, ...remaining } = unitList;
+    setUnitList(remaining);
+  };
+
+  const addUpgrade = (unitKey, upgradeKey, upgrade) => {
+      let unitListCpy = {...unitList};
+      unitListCpy[unitKey].selected_upgrades[upgradeKey] = upgrade;
+      setUnitList(unitListCpy);
```

Comment:
You're still mutating state since the content of `unitList` is still the actual state values!
```suggestion
      setUnitList({
          ...unitList,
          [unitKey]: {
              ...unitList[unitKey],
              selected_upgrades: {
                  ...unitList[unitKey].selected_upgrades,
                  [upgradeKey]: upgrade
              }
          }
      });
```

Reply:
Yip, this would be a the way to avoid mutations.

It does feel a bit messy though since you have to spread over so many objects. One of the downsides of having such a deeply nested state structure. This does have its own trade-offs which I'll mention in a sec but one option if this starts feeling unmanageable is to actually flatten the state. ie: have a separate state value for selected upgrades and unit list. selected upgrades would be a separate list with the unit key as the keys in the map.

Pros:
If you tend to update upgrades on their own a lot, you can avoid touching any irrelevant state instead of copying over the whole unit object over and over.

Cons:
Actions that remove the unit entirely are a bit more difficult to manage because you have to remove the references in each smaller piece of state.

Not saying that you need to do this here or now, but your unit list state is only going to get bigger and more complex over time so it's worth thinking about how you're going to maintain/manage it.